File: primat/qed_pressure.py
# -*- coding: utf-8 -*-
"""
qed_pressure.py — Analytical computation of QED plasma-pressure corrections
============================================================================

Computes the finite-temperature QED interaction-pressure corrections
δP(T), dδP/dT, and d²δP/dT² that enter the EM plasma thermodynamics
during BBN.  These supplement the free ideal-gas photon + e± expressions
with the leading electromagnetic interactions.

Physical background
-------------------
The QED interaction pressure is a finite-temperature correction arising
from the QED interaction between photons and electrons in the hot plasma.
It is decomposed into three contributions in increasing order of the
electromagnetic coupling e (α = e²/(4π)):

  δP = δP_a [O(e²)]  +  δP_{e3} [O(e³)]  +  δP_b [O(e⁴)]

where (following PRIMAT-Main.m and Phys. Rep. §II.E):

  δP_a(T)  = (α/π) T⁴ [-(2/3) I₀₁(x) - (2/π²) I₀₁(x)²]
              Leading O(α) one-loop correction (Frenkel–Galitskii–Migdal).

  δP_{e3}(T) = α^{3/2} (4/3)√(2π) T⁴ [(I₀₁(x)+I₂₋₁(x))/π²]^{3/2}
               O(α^{3/2}) ring/plasmon contribution (Blaizot–Zinn-Justin).

  δP_b(T)  = T⁴ ∫₀^∞ ∫₀^∞ F(p₁,p₂,x) dp₁ dp₂           [O(α²), optional]
             F = (α/π³) x² p₁ p₂ / (e₁ e₂)
                 × ln|(p₁+p₂)/(p₁-p₂)| / ((e^{e₁}+1)(e^{e₂}+1))

Here x = mₑ/T (dimensionless), eᵢ = √(pᵢ²+x²), and:

  I₀₁(x) = ∫₀^∞ p² / [√(p²+x²)(e^{√(p²+x²)}+1)] dp
           = ∫_x^∞ √(E²−x²)/(e^E+1) dE           (PRIMAT: Imn[1][0,1][x])

  I₂₋₁(x) = ∫₀^∞ √(p²+x²) / (e^{√(p²+x²)}+1) dp
            = ∫_x^∞ E²/[√(E²−x²)(e^E+1)] dE      (PRIMAT: Imn[1][2,-1][x])

The dominant term is δP_a, which is negative (interaction reduces the
pressure relative to the ideal gas).  The ring term δP_{e3} is positive
and roughly 10× smaller.  The two-loop exchange δP_b is typically 100×
smaller still and is optional.

File format
-----------
The results are stored in ``rates/plasma/QED_*.txt`` with three columns:
  T [MeV]  |  quantity_e2  |  quantity_e3

where ``_e2`` = δP_a (order e²) and ``_e3`` = δP_{e3} (order e³).
When loaded by :mod:`primat.plasma`, both columns are summed to give the
total correction.  The δP_b term would require a separate flag and file.

Usage
-----
>>> from primat.qed_pressure import compute_qed_pressure_tables, save_qed_tables
>>> tables = compute_qed_pressure_tables()  # ~0.3 s on a modern laptop
>>> save_qed_tables(tables, "/path/to/data/plasma/")

Reference
---------
Pitrou, Coc, Uzan & Vangioni, Phys. Rep. 2018 (arXiv:1806.11095), §II.E
PRIMAT-Main.m: ``dPa``, ``dPe3``, ``dPb`` definitions (lines 920, 939, 949)
"""

import os
import numpy as np
from scipy.integrate import quad, dblquad
from scipy.interpolate import CubicSpline

# Physical constants — kept local to this module (not imported from config)
# to allow standalone use in generate_rates/ scripts.
_ALPHA_FS = 1. / 137.035999084   # fine-structure constant (CODATA 2018)
_ME_MEV   = 0.5109989461         # electron mass [MeV]

# Low-x cutoff: for x = mₑ/T > 50 (T < mₑ/50 ≈ 10 keV) the e± are so
# non-relativistic that δP is effectively zero (Boltzmann-suppressed).
_X_NONREL_CUTOFF = 50.

# Upper limit for 1D momentum integrals, in units of x = mₑ/T.  The
# integrand decays as e^{-p} for large p, so p_max = 500 is more than
# sufficient even at very high temperatures.
_P_UPPER = 500.


# ---------------------------------------------------------------------------
# Fermi-Dirac momentum integrals I₀₁ and I₂₋₁
# ---------------------------------------------------------------------------

def _I01(x):
    """Fermi-Dirac phase-space integral I₀₁(x) [dimensionless].

    Defined as (PRIMAT: Imn[1][0,1][x]):

        I₀₁(x) = ∫₀^∞ p² / [√(p²+x²)(e^{√(p²+x²)}+1)] dp

    Equivalently (change of variable E = √(p²+x²)):

        I₀₁(x) = ∫_x^∞ √(E²−x²) / (e^E+1) dE

    The p-space form is used here because it is non-singular at the lower
    limit, making scipy.quad straightforward to apply.

    Parameters
    ----------
    x : float
        Dimensionless ratio mₑ/T.

    Returns
    -------
    float
        I₀₁(x) in natural units (ℏ = c = kB = 1).

    Example
    -------
    >>> _I01(0.0)   # ultra-relativistic limit → π²/12 ≈ 0.822
    >>> _I01(0.5)   # semi-relativistic (T ~ 1 MeV)
    """
    if x > _X_NONREL_CUTOFF:
        return 0.
    def integrand(p):
        E = np.sqrt(p * p + x * x)
        return p * p / (E * (np.exp(E) + 1.))
    result, _ = quad(integrand, 0., _P_UPPER,
                     epsabs=1e-13, epsrel=1e-13, limit=300)
    return result


def _I2m1(x):
    """Fermi-Dirac phase-space integral I₂₋₁(x) [dimensionless].

    Defined as (PRIMAT: Imn[1][2,-1][x]):

        I₂₋₁(x) = ∫₀^∞ √(p²+x²) / (e^{√(p²+x²)}+1) dp

    Equivalently:

        I₂₋₁(x) = ∫_x^∞ E² / [√(E²−x²)(e^E+1)] dE

    The p-space form removes the 1/√(E²−x²) singularity at the lower
    limit, so no special handling is needed.

    Parameters
    ----------
    x : float
        Dimensionless ratio mₑ/T.

    Returns
    -------
    float
        I₂₋₁(x) in natural units.

    Example
    -------
    >>> _I2m1(0.0)   # ultra-relativistic limit → π²/12 ≈ 0.822
    """
    if x > _X_NONREL_CUTOFF:
        return 0.
    def integrand(p):
        E = np.sqrt(p * p + x * x)
        return E / (np.exp(E) + 1.)
    result, _ = quad(integrand, 0., _P_UPPER,
                     epsabs=1e-13, epsrel=1e-13, limit=300)
    return result


# ---------------------------------------------------------------------------
# Three contributions to δP
# ---------------------------------------------------------------------------

def _dPa(T, alpha=_ALPHA_FS, me=_ME_MEV):
    """O(e²) QED interaction-pressure correction δP_a(T) [MeV⁴].

    The leading one-loop correction to the electromagnetic plasma pressure
    from the QED interaction between photons and electrons
    (PRIMAT: ``dPa``; Phys. Rep. §II.E):

        δP_a = (α/π) T⁴ [−(2/3) I₀₁(x) − (2/π²) I₀₁(x)²]

    This is negative (interaction lowers the pressure) and is the dominant
    QED correction, of order α ∼ 7×10⁻³.

    Parameters
    ----------
    T : float
        Photon temperature [MeV].
    alpha : float, optional
        Fine-structure constant (default: 1/137.035999084).
    me : float, optional
        Electron mass [MeV] (default: 0.5109989461).

    Returns
    -------
    float
        δP_a in MeV⁴.

    Example
    -------
    >>> _dPa(10.0)   # at T = 10 MeV
    """
    x = me / T
    I01 = _I01(x)
    return alpha / np.pi * T**4 * (-2./3. * I01 - 2./np.pi**2 * I01**2)


def _dPe3(T, alpha=_ALPHA_FS, me=_ME_MEV):
    """O(e³) QED interaction-pressure correction δP_{e3}(T) [MeV⁴].

    The O(α^{3/2}) ring/plasmon contribution to the QED pressure,
    arising from collective plasma oscillations (PRIMAT: ``dPe3``):

        δP_{e3} = α^{3/2} (4/3)√(2π) T⁴ [(I₀₁+I₂₋₁)/π²]^{3/2}

    This is positive and roughly 10× smaller than δP_a.

    Parameters
    ----------
    T : float
        Photon temperature [MeV].
    alpha : float, optional
        Fine-structure constant.
    me : float, optional
        Electron mass [MeV].

    Returns
    -------
    float
        δP_{e3} in MeV⁴.

    Example
    -------
    >>> _dPe3(10.0)   # at T = 10 MeV
    """
    x = me / T
    I01  = _I01(x)
    I2m1 = _I2m1(x)
    combo = (I01 + I2m1) / np.pi**2
    if combo <= 0.:
        return 0.
    return alpha**(3./2.) * (4./3.) * np.sqrt(2. * np.pi) * T**4 * combo**(3./2.)


def _dPb(T, alpha=_ALPHA_FS, me=_ME_MEV, epsrel=1e-4):
    """O(e⁴) QED interaction-pressure correction δP_b(T) [MeV⁴].

    The two-loop exchange contribution, corresponding to PRIMAT's
    ``dPb`` (``$CompleteQEDPressure=True``):

        δP_b = T⁴ ∫₀^∞ ∫₀^∞ F(p₁,p₂,x) dp₁ dp₂

    with

        F(p₁,p₂,x) = (α/π³) x² p₁ p₂ / (e₁ e₂)
                      × ln|(p₁+p₂)/(p₁−p₂)| / ((e^{e₁}+1)(e^{e₂}+1))

    where eᵢ = √(pᵢ²+x²).  The integrand is symmetric in p₁↔p₂, and the
    logarithm has an integrable singularity at p₁ = p₂.

    **Note**: this term is O(α²) ≈ 5×10⁻⁵ and is NOT included in the
    standard primat QED tables (which only store δP_a + δP_{e3}).
    It is provided here for completeness.  Computing it is expensive
    (~10–60 s per temperature point at low precision).

    Parameters
    ----------
    T : float
        Photon temperature [MeV].
    alpha : float, optional
        Fine-structure constant.
    me : float, optional
        Electron mass [MeV].
    epsrel : float, optional
        Relative accuracy target for the 2D numerical integration
        (default 1e-4, matching PRIMAT's ``PrecisionGoal->4``).

    Returns
    -------
    float
        δP_b in MeV⁴.
    """
    x = me / T
    if x > _X_NONREL_CUTOFF:
        return 0.

    # Upper momentum limit: at least 20, or 20x (non-relativistic: pmax ~ x)
    p_upper = max(20., 20. * x)

    def integrand(p2, p1):
        # eᵢ = √(pᵢ² + x²); the log factor is regularised by treating the
        # p1 == p2 singularity as integrable (verified: log divergence, area 0)
        e1 = np.sqrt(p1 * p1 + x * x)
        e2 = np.sqrt(p2 * p2 + x * x)
        if abs(p1 - p2) < 1e-14 * (p1 + p2 + 1e-10):
            return 0.   # contribution zero on the diagonal p1=p2
        log_factor = np.log(abs((p1 + p2) / (p1 - p2)))
        fd1 = 1. / (np.exp(e1) + 1.)
        fd2 = 1. / (np.exp(e2) + 1.)
        return (alpha / np.pi**3) * x**2 * p1 * p2 / (e1 * e2) * log_factor * fd1 * fd2

    result, _ = dblquad(integrand, 0., p_upper,
                        lambda p1: 0., lambda p1: p_upper,
                        epsrel=epsrel, epsabs=0.)
    return T**4 * result


# ---------------------------------------------------------------------------
# Grid computation and file I/O
# ---------------------------------------------------------------------------

def compute_qed_pressure_tables(T_min=1e-3, T_max=1e2, n_pts=500,
                                alpha=_ALPHA_FS, me=_ME_MEV,
                                include_dPb=False, verbose=True):
    """Compute δP, dδP/dT, d²δP/dT² on a temperature grid [MeV].

    Evaluates the O(e²) and O(e³) QED corrections to the EM plasma
    pressure (and optionally the O(e⁴) exchange term) on a log-spaced
    temperature grid, then differentiates numerically using a cubic spline.

    The two-column format matches the files loaded by
    :func:`primat.plasma._load_tables`:
      column 0 = T [MeV]
      column 1 = δP_a(T)   [MeV⁴]  (O(e²) = O(α))
      column 2 = δP_{e3}(T) [MeV⁴] (O(e³) = O(α^{3/2}))

    When ``include_dPb=True`` a third column for δP_b is added to the
    ``dP`` table, and the derivatives are recomputed accordingly.

    Parameters
    ----------
    T_min : float
        Minimum temperature [MeV] (default 1e-3, well below e± freeze-out).
    T_max : float
        Maximum temperature [MeV] (default 100, well above BBN start).
    n_pts : int
        Number of log-spaced temperature grid points (default 500).
    alpha : float
        Fine-structure constant (default: CODATA 2018 value).
    me : float
        Electron mass [MeV] (default: 0.5109989461).
    include_dPb : bool
        If True, also compute the expensive O(e⁴) two-loop term δP_b
        (adds ~10–60 s per temperature point; default False).
    verbose : bool
        Print progress messages (default True).

    Returns
    -------
    dict
        Keys: ``"T"``, ``"dP_e2"``, ``"dP_e3"``, ``"d_dP_e2_dT"``,
        ``"d_dP_e3_dT"``, ``"d2_dP_e2_dT2"``, ``"d2_dP_e3_dT2"``,
        and optionally ``"dP_b"``, ``"d_dPb_dT"``, ``"d2_dPb_dT2"``.
        All arrays have length ``n_pts``.

    Notes
    -----
    The derivatives dδP/dT and d²δP/dT² are obtained from a CubicSpline
    fit to the tabulated δP values, not from analytic differentiation of
    the Fermi-Dirac integrals.  The analytic route would require four
    additional quadratures per temperature point and would be ~7× slower
    with no practical accuracy gain: the spline derivatives agree with
    direct finite differences on _dPa/_dPe3 to <0.01% at all T.

    Example
    -------
    >>> tables = compute_qed_pressure_tables(n_pts=100, verbose=False)
    >>> tables["T"].shape
    (100,)
    """
    T_grid = np.logspace(np.log10(T_min), np.log10(T_max), n_pts)
    dP_e2  = np.zeros(n_pts)
    dP_e3  = np.zeros(n_pts)
    dP_b   = np.zeros(n_pts) if include_dPb else None

    for i, T in enumerate(T_grid):
        if verbose and i % max(1, n_pts // 10) == 0:
            print(f"  [QED] Computing T = {T:.3e} MeV  ({i+1}/{n_pts})")
        dP_e2[i] = _dPa(T, alpha=alpha, me=me)
        dP_e3[i] = _dPe3(T, alpha=alpha, me=me)
        if include_dPb:
            dP_b[i] = _dPb(T, alpha=alpha, me=me)

    # Differentiate numerically using a cubic spline; this avoids having to
    # differentiate the integrands analytically.
    spl_e2 = CubicSpline(T_grid, dP_e2)
    spl_e3 = CubicSpline(T_grid, dP_e3)
    d_e2   = spl_e2(T_grid, 1)   # first derivative
    d2_e2  = spl_e2(T_grid, 2)   # second derivative
    d_e3   = spl_e3(T_grid, 1)
    d2_e3  = spl_e3(T_grid, 2)

    out = {"T": T_grid,
           "dP_e2": dP_e2, "dP_e3": dP_e3,
           "d_dP_e2_dT": d_e2, "d_dP_e3_dT": d_e3,
           "d2_dP_e2_dT2": d2_e2, "d2_dP_e3_dT2": d2_e3}

    if include_dPb:
        spl_b  = CubicSpline(T_grid, dP_b)
        out["dP_b"]        = dP_b
        out["d_dPb_dT"]    = spl_b(T_grid, 1)
        out["d2_dPb_dT2"]  = spl_b(T_grid, 2)

    return out


def save_qed_tables(tables, plasma_dir, verbose=True):
    """Write the computed QED tables to ``rates/plasma/*.txt`` files.

    Produces three files whose format matches the ones loaded by
    :func:`primat.plasma._load_tables`:

      - ``QED_P_int.txt``      — δP (columns: T, δP_a, δP_{e3})
      - ``QED_dP_intdT.txt``   — dδP/dT (columns: T, dδP_a/dT, dδP_{e3}/dT)
      - ``QED_d2P_intdT2.txt`` — d²δP/dT² (similar)

    Parameters
    ----------
    tables : dict
        Output of :func:`compute_qed_pressure_tables`.
    plasma_dir : str
        Path to the ``rates/plasma/`` directory.
    verbose : bool
        Print confirmation messages (default True).

    Example
    -------
    >>> save_qed_tables(tables, "rates/plasma/")
    """
    T    = tables["T"]
    e2   = tables["dP_e2"]
    e3   = tables["dP_e3"]
    de2  = tables["d_dP_e2_dT"]
    de3  = tables["d_dP_e3_dT"]
    d2e2 = tables["d2_dP_e2_dT2"]
    d2e3 = tables["d2_dP_e3_dT2"]

    hdr_P   = ("Source: primat qed_pressure.py — computed from PRIMAT formulas\n"
               "T (MeV)           P_int (e^2)          P_int (e^3)")
    hdr_dP  = ("Source: primat qed_pressure.py — computed from PRIMAT formulas\n"
               "T (MeV)           dP_int/dT (e^2)      dP_int/dT (e^3)")
    hdr_d2P = ("Source: primat qed_pressure.py — computed from PRIMAT formulas\n"
               "T (MeV)           d2P_int/dT2 (e^2)    d2P_int/dT2 (e^3)")

    fmt = "%.6E"
    np.savetxt(os.path.join(plasma_dir, "QED_P_int.txt"),
               np.column_stack([T, e2, e3]), header=hdr_P, fmt=fmt)
    np.savetxt(os.path.join(plasma_dir, "QED_dP_intdT.txt"),
               np.column_stack([T, de2, de3]), header=hdr_dP, fmt=fmt)
    np.savetxt(os.path.join(plasma_dir, "QED_d2P_intdT2.txt"),
               np.column_stack([T, d2e2, d2e3]), header=hdr_d2P, fmt=fmt)

    if verbose:
        print(f"[QED]  Tables written to {plasma_dir}:")
        print(f"       QED_P_int.txt, QED_dP_intdT.txt, QED_d2P_intdT2.txt")
        print(f"       T range: {T[0]:.3e}–{T[-1]:.3e} MeV  ({len(T)} points)")

File: primat-c/src/qed_pressure.c
/* qed_pressure.c -- see qed_pressure.h. */
#include "qed_pressure.h"
#include "quad.h"
#include "spline.h"

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Mirrors qed_pressure.py's module-level constants exactly (see that
 * file's comments for the rationale of each). */
#define X_NONREL_CUTOFF 50.0  /* x = me/T above which dP is set to 0 */
#define P_UPPER         500.0 /* upper momentum-integration limit */
#define QUAD_TOL        1e-13 /* matches scipy.quad's epsabs=epsrel=1e-13 */
#define QUAD_MAX_DEPTH  30

static double i01_integrand(double p, void *ctx)
{
    double x = *(double *)ctx;
    double E = sqrt(p * p + x * x);
    /* At p=x=0 both numerator and denominator vanish (0/0); the true
     * limit is 0 (numerator ~ p^2, denominator ~ p, so the ratio ~ p -> 0).
     * scipy.quad never samples this exact point (Gauss-Kronrod nodes
     * exclude the interval endpoints), so qed_pressure.py never hits it;
     * cpr_quad_adaptive's Simpson rule does sample the endpoint, so this
     * guard is needed for the x=0 case to avoid a NaN. */
    if (E == 0.0) return 0.0;
    return p * p / (E * (exp(E) + 1.0));
}

static double i2m1_integrand(double p, void *ctx)
{
    double x = *(double *)ctx;
    double E = sqrt(p * p + x * x);
    return E / (exp(E) + 1.0);
}

/* Fixed breakpoints, geometrically widening from 0 up to P_UPPER. Both
 * integrands here are unimodal Boltzmann-like bumps p^a exp(-sqrt(p^2+x^2))
 * whose peak location grows with x (roughly sqrt(2x) for x >~ few, by
 * maximising p^2 exp(-sqrt(p^2+x^2))); for the largest x this module ever
 * evaluates (the X_NONREL_CUTOFF=50 cutoff above), the peak sits around
 * p ~ 10. A single cpr_quad_adaptive call across [0, P_UPPER] with P_UPPER
 * = 500 fails here: its first-level Simpson sample sits at {0, 250, 500},
 * entirely past the peak, where the integrand has already underflowed to
 * 0 -- coarse and refined estimates then agree (both ~0) and the
 * recursion accepts a result of 0, silently missing the entire peak
 * (verified: this reproduces exactly as I01(0) -> 0 instead of pi^2/12).
 * Splitting into pre-defined sub-intervals densest near 0 guarantees every
 * sub-interval's own first-level sample lands inside or adjacent to the
 * region where the integrand actually varies, for any x in [0, 50]. */
static const double BREAKPOINTS[] = {
    0.0, 0.5, 1.0, 2.0, 4.0, 8.0, 16.0, 32.0, 64.0, 128.0, 256.0, P_UPPER
};
#define N_BREAKPOINTS (sizeof(BREAKPOINTS) / sizeof(BREAKPOINTS[0]))

static double integrate_segmented(CPRQuadFunc f, void *ctx)
{
    double total = 0.0;
    double seg_tol = QUAD_TOL / (double)(N_BREAKPOINTS - 1);
    for (size_t i = 0; i + 1 < N_BREAKPOINTS; i++)
        total += cpr_quad_adaptive(f, ctx, BREAKPOINTS[i], BREAKPOINTS[i + 1],
                                    seg_tol, QUAD_MAX_DEPTH, NULL);
    return total;
}

double cpr_qed_I01(double x)
{
    if (x > X_NONREL_CUTOFF) return 0.0;
    return integrate_segmented(i01_integrand, &x);
}

double cpr_qed_I2m1(double x)
{
    if (x > X_NONREL_CUTOFF) return 0.0;
    return integrate_segmented(i2m1_integrand, &x);
}

double cpr_qed_dPa(double T, double alpha, double me)
{
    double x = me / T;
    double I01 = cpr_qed_I01(x);
    return alpha / M_PI * pow(T, 4.0) * (-2.0 / 3.0 * I01 - 2.0 / (M_PI * M_PI) * I01 * I01);
}

double cpr_qed_dPe3(double T, double alpha, double me)
{
    double x = me / T;
    double I01  = cpr_qed_I01(x);
    double I2m1 = cpr_qed_I2m1(x);
    double combo = (I01 + I2m1) / (M_PI * M_PI);
    if (combo <= 0.0) return 0.0;
    return pow(alpha, 1.5) * (4.0 / 3.0) * sqrt(2.0 * M_PI) * pow(T, 4.0) * pow(combo, 1.5);
}

int cpr_qed_compute_tables(double T_min, double T_max, size_t n_pts,
                            double alpha, double me,
                            CPRQEDTables *out, char **errmsg)
{
    if (n_pts < 4) {
        *errmsg = strdup("cpr_qed_compute_tables: n_pts must be >= 4 (not-a-knot spline minimum)");
        return 1;
    }

    memset(out, 0, sizeof(*out));
    out->n = n_pts;
    out->T              = malloc(n_pts * sizeof(double));
    out->dP_e2           = malloc(n_pts * sizeof(double));
    out->dP_e3           = malloc(n_pts * sizeof(double));
    out->d_dP_e2_dT      = malloc(n_pts * sizeof(double));
    out->d_dP_e3_dT      = malloc(n_pts * sizeof(double));
    out->d2_dP_e2_dT2    = malloc(n_pts * sizeof(double));
    out->d2_dP_e3_dT2    = malloc(n_pts * sizeof(double));

    /* Log-spaced grid (np.logspace(log10(T_min), log10(T_max), n_pts)). */
    double log_min = log10(T_min), log_max = log10(T_max);
    for (size_t i = 0; i < n_pts; i++) {
        double frac = (n_pts == 1) ? 0.0 : (double)i / (double)(n_pts - 1);
        out->T[i] = pow(10.0, log_min + frac * (log_max - log_min));
        out->dP_e2[i] = cpr_qed_dPa(out->T[i], alpha, me);
        out->dP_e3[i] = cpr_qed_dPe3(out->T[i], alpha, me);
    }

    /* Differentiate numerically via not-a-knot cubic splines (matches
     * scipy.interpolate.CubicSpline's default boundary condition, used by
     * compute_qed_pressure_tables). */
    CPRCubicSpline spl_e2, spl_e3;
    if (cpr_cubic_spline_fit_notaknot(out->T, out->dP_e2, n_pts, &spl_e2, errmsg)) {
        cpr_qed_tables_free(out);
        return 1;
    }
    if (cpr_cubic_spline_fit_notaknot(out->T, out->dP_e3, n_pts, &spl_e3, errmsg)) {
        cpr_cubic_spline_free(&spl_e2);
        cpr_qed_tables_free(out);
        return 1;
    }
    for (size_t i = 0; i < n_pts; i++) {
        /* CPRCubicSpline stores per-segment polynomial coefficients
         * directly (b = 1st deriv, 2c = 2nd deriv, at the left knot of
         * each segment); evaluate analytically at each knot rather than
         * via cpr_cubic_spline_eval (which is the y(x) evaluator, not a
         * derivative one) by locating the segment whose left knot is T[i]
         * (every grid point is itself a knot, so dx=0 there and the
         * derivatives are exactly b[i] and 2*c[i] -- except at the last
         * knot, which belongs to segment n-2). */
        size_t seg = (i < n_pts - 1) ? i : n_pts - 2;
        double dx = out->T[i] - spl_e2.x[seg];
        out->d_dP_e2_dT[i]   = spl_e2.b[seg] + 2.0 * spl_e2.c[seg] * dx + 3.0 * spl_e2.d[seg] * dx * dx;
        out->d2_dP_e2_dT2[i] = 2.0 * spl_e2.c[seg] + 6.0 * spl_e2.d[seg] * dx;
        dx = out->T[i] - spl_e3.x[seg];
        out->d_dP_e3_dT[i]   = spl_e3.b[seg] + 2.0 * spl_e3.c[seg] * dx + 3.0 * spl_e3.d[seg] * dx * dx;
        out->d2_dP_e3_dT2[i] = 2.0 * spl_e3.c[seg] + 6.0 * spl_e3.d[seg] * dx;
    }
    cpr_cubic_spline_free(&spl_e2);
    cpr_cubic_spline_free(&spl_e3);
    return 0;
}

void cpr_qed_tables_free(CPRQEDTables *t)
{
    free(t->T); free(t->dP_e2); free(t->dP_e3);
    free(t->d_dP_e2_dT); free(t->d_dP_e3_dT);
    free(t->d2_dP_e2_dT2); free(t->d2_dP_e3_dT2);
    memset(t, 0, sizeof(*t));
}

/* Writes two 4-column files, one per order in e, matching
 * numpy.savetxt(fmt="%.6E") output from Python's save_qed_tables. Each
 * file's columns are: T [MeV]  dP [MeV^4]  d(dP)/dT [MeV^3]  d2(dP)/dT2
 * [MeV^2]. All '#'-prefixed header lines are skipped by cpr_table_read. */
static int write_one_qed_file(const char *path, const char *src_tag,
                               const char *phys_tag, const char *ref_tag,
                               const char *col_hdr,
                               const double *T, const double *dP,
                               const double *ddP, const double *d2dP,
                               size_t n, char **errmsg)
{
    FILE *fp = fopen(path, "w");
    if (!fp) {
        char buf[512];
        snprintf(buf, sizeof(buf), "cpr_qed_save_tables: cannot open %s for writing", path);
        *errmsg = strdup(buf);
        return 1;
    }
    fprintf(fp, "# Source: %s\n", src_tag);
    fprintf(fp, "# %s\n", phys_tag);
    fprintf(fp, "# Reference: %s\n", ref_tag);
    fprintf(fp, "# %s\n", col_hdr);
    for (size_t i = 0; i < n; i++)
        fprintf(fp, "%.6E %.6E %.6E %.6E\n", T[i], dP[i], ddP[i], d2dP[i]);
    fclose(fp);
    return 0;
}

int cpr_qed_save_tables(const CPRQEDTables *t, const char *plasma_dir, char **errmsg)
{
    char path_e2[1024], path_e3[1024];
    snprintf(path_e2, sizeof(path_e2), "%s/QED_pressure_correction_e2.txt", plasma_dir);
    snprintf(path_e3, sizeof(path_e3), "%s/QED_pressure_correction_e3.txt", plasma_dir);

    if (write_one_qed_file(path_e2,
            "CPRIMAT qed_pressure.c -- QED plasma-pressure correction delta_P_a(T)",
            "delta_P_a: O(e^2), one-loop (Frenkel-Galitskii-Migdal)",
            "Pitrou et al., Phys. Rep. (2018), eq. 47; PRIMAT-Main.m: dPa",
            "T [MeV]       dP_a [MeV^4]      d(dP_a)/dT [MeV^3]  d2(dP_a)/dT2 [MeV^2]",
            t->T, t->dP_e2, t->d_dP_e2_dT, t->d2_dP_e2_dT2, t->n, errmsg))
        return 1;

    if (write_one_qed_file(path_e3,
            "CPRIMAT qed_pressure.c -- QED plasma-pressure correction delta_P_e3(T)",
            "delta_P_e3: O(e^3), ring/plasmon (Blaizot-Zinn-Justin)",
            "Pitrou et al., Phys. Rep. (2018), eq. 47; PRIMAT-Main.m: dPe3",
            "T [MeV]       dP_e3 [MeV^4]     d(dP_e3)/dT [MeV^3]  d2(dP_e3)/dT2 [MeV^2]",
            t->T, t->dP_e3, t->d_dP_e3_dT, t->d2_dP_e3_dT2, t->n, errmsg))
        return 1;

    return 0;
}
